### PR TITLE
Broadcast epoch block

### PR DESCRIPTION
**Pull request by Frozen**
Broadcasts last epoch block from shard 0 to shard 1.

### CODE REVIEW

**Comment by Frozen** [comment]
@sophoah

**Comment by mur-me** [comment]
I will test this version against devnet

**Comment by mur-me** [comment]
Hey @Frozen, 

I've tried your PR several times and all the times shard 1 consensus was broken around 12 or 15 epoch.
---
To reproduce:
1. I've used your PR branch + two cherrypicks for:
* logs aggregation  - https://github.com/harmony-one/harmony/pull/4758
* crosslinks fix -  https://github.com/harmony-one/harmony/pull/4751, yes there is other one from Gheis, but it is still in progress
```
git cherry-pick  8b8dcd8ffb814309aea66a4ccce05f41b629cefa
git cherry-pick 4df7ea82450f2f2864d93dfcede60f6cc9b67dd0
```
2. Start the localnet with `make debug-ext`
3. Start the logs aggregation `make debug-start-log-aggregator`
4. Wait until the 15th epoch

Result:
* one of external validators have stopped to work on the 9 epoch - `"last-epoch-in-committee": 8` and have a lot of:
``` json
{
  "level": "error",
  "port": "9116",
  "ip": "127.0.0.1",
  "shardID": 1,
  "myBlock": 736,
  "myViewID": 742,
  "phase": "Announce",
  "mode": "Listening",
  "error": "no elected bls keys provided",
  "message-type": "PREPARE",
  "caller": "/home/uladzislau/go/src/github.com/harmony-one/harmony/consensus/validator.go:436",
  "time": "2024-09-18T20:04:04.066683495+03:00",
  "message": "could not construct message"
}
```
* shard 1 stopped completely on the 15->16 epoch, all the node are in the infinite ViewChanging mode

Logs have a lot of:
``` json
{
  "level": "error",
  "port": "9104",
  "ip": "127.0.0.1",
  "shardID": 1,
  "myBlock": 682,
  "myViewID": 688,
  "phase": "Prepare",
  "mode": "Normal",
  "error": "Block verification failed: Block verification failed: [VerifyNewBlock] Cannot Verify New Block!!! block-hash 0xb9c121db4d58856676fefda069d0881b7f15594815ca5697d52ed8fb5edb6fe6 txn-count 0: [ValidateNewBlock] Cannot verify shard state for the new block: failed to read shard state from DB%!(EXTRA string=epoch: %d, uint64=14)",
  "MsgBlockNum": 682,
  "MsgViewID": 688,
  "caller": "/home/uladzislau/go/src/github.com/harmony-one/harmony/consensus/validator.go:243",
  "time": "2024-09-18T20:02:09.223556278+03:00",
  "message": "[OnPrepared] failed to verify new block"
}
```
and
``` json
{
  "level": "error",
  "port": "9102",
  "ip": "127.0.0.1",
  "shardID": 1,
  "myBlock": 682,
  "myViewID": 688,
  "phase": "Announce",
  "mode": "Syncing",
  "error": "Block verification failed: Block verification failed: [VerifyNewBlock] Cannot Verify New Block!!! block-hash 0x7d5b18f75c15fac829374509d536aaa05e0ac94190d37207704655720a3de745 txn-count 0: unknown ancestor",
  "MsgBlockNum": 686,
  "MsgViewID": 692,
  "caller": "/home/uladzislau/go/src/github.com/harmony-one/harmony/consensus/validator.go:243",
  "time": "2024-09-18T20:02:23.481702369+03:00",
  "message": "[OnPrepared] failed to verify new block"
}
```

**Comment by sophoah** [comment]
@Frozen #4754  (crosslink fix) has been merged. Let's rebase that PR so we don't have to cherry pick in order to test that one

**Comment by sophoah** [comment]
hey @Frozen, is there any potential conflict with epoch sync download with this PR ?

**Comment by Frozen** [comment]
> hey @Frozen, is there any potential conflict with epoch sync download with this PR ?

barely possible

**Review by GheisMohammadi — APPROVED** [review]

**Comment by sophoah** [comment]
> barely possible

could you explain why ?

File: consensus/consensus_v2.go [hunk 1]
```diff
@@ -184,7 +186,7 @@ func (consensus *Consensus) finalCommit() {
 	// Note: leader already sent 67% commit in preCommit. The 100% commit won't be sent immediately
 	// to save network traffic. It will only be sent in retry if consensus doesn't move forward.
 	// Or if the leader is changed for next block, the 100% committed sig will be sent to the next leader immediately.
-	if !consensus.isLeader() || block.IsLastBlockInEpoch() {
+	if !isLeader || block.IsLastBlockInEpoch() {
```
Comment (sophoah) [inline consensus/consensus_v2.go:189]: since isLeader() is already in consensus, why passing isLeader to the function now ?

Comment (Frozen) [reply consensus/consensus_v2.go:189]: Function runs in goroutine and sometimes `consensus.isLeader()` returns `false` where should be `true`

**Review by sophoah — APPROVED** [review]

**Comment by mur-me** [comment]
I've done several tests on the devnet:
1. one of shard 1 validator nodes has been successfully synced from 0, beacon chain as the 1st thing, and shard 1 DB afterwards
2.  tried to mix nodes with this fix and without it and have 2 situations and both of them aren't critical:
* shard 0 leader is new release -> **problem occurs time  to time, not in every epoch change**, but if it occurs, it will be 1 view change 
* shard 0 leader is old release -> it will be 2 view changes 

Why this isn't critical:
* epoch change occurs once in 18.2 hour in the mainnet
* we will have this problem less and less after validators start to updating to the new release

**Pull request merged**